modules/reactor/doc/content/source/meshgenerators/AssemblyMeshGenerator.md

```diff
@@ -0,0 +1,43 @@
+This object automates the use and functionality of the [`CartesianIDPatternedMeshGenerator`](CartesianIDPatternedMeshGenerator.md) for cartesian  reactor geometry, [`HexIDPatternedMeshGenerator`](HexIDPatternedMeshGenerator.md) for hexagonal reactor geometry and, if extruding to three dimension, the [`FancyExtruderGenerator'](FancyExtruderGenerator.md) through the use of the `MeshSubgenerator` functionality and supporting functionality from [`RenameBoundaryGenerator`](RenameBoundaryGenerator.md) and [`PlaneIDMeshGenerator'](PlaneIDMeshGenerator.md). In addition to the functionality of the `MeshGenerators` used, this object also automates block ID assignment for background and duct regions (only available for hexagonal assemblies) and boundary ID and name assignment. 
+The `AssemblyMeshGenerator` object adopts much of the existing input structure of patterned MeshGenerators but but adopts the use of parameters that are more typical for reactor design. 
+
+## Block ID Information
+The [!param](/Mesh/AssemblyMeshGenerator/background_region_id) and [!param](/Mesh/AssemblyMeshGenerator/duct_region_ids) parameters are used to identify regions within the assembly, that do not belong to one of the constituent pins, that belong to the same block both radially and axially. This functionality is intended for easy identification of regions within the mesh that will have the same properties, such as material assignments, and will assign the region ID to both the subdomain (block) ID and name.
+The user defined ID assignment using [!param](/Mesh/AssemblyMeshGenerator/background_region_id) and [!param](/Mesh/AssemblyMeshGenerator/duct_region_ids) should be given as a single value per axial layer for the background region and a vector, starting from the inner-most duct region, of the IDs for the appropriate axial layer starting from the bottom of the geometry. 
+The automated ID assignment, via the [!param](/Mesh/ReactorMeshParams/procedural_region_ids) option, concatenates the [!param](/Mesh/AssemblyMeshGenerator/assembly_type) and the peripheral radial mesh region number (indexed starting at 0 for the background region) with the maximum number of zeroes separating the two in order to differentiate the regions from regions in pins. In the case that the concatenatation exceeds the type limit, the lowest digits of the [!param](/Mesh/AssemblyMeshGenerator/assembly_type) will be truncated. The same ID will be used for all axial layers in three dimensional meshes.
+
+For example a hexagonal assembly with a type ID of 37 and two ducts would procedurally generate three region IDs:
+- A background region ID of 37000
+- A first duct region ID of 37001
+- A second duct region ID of 37002
+
+If the same assembly instead has a type ID of 65535 the same three regions would be assigned as:
+- A background region ID of 65530
+- A first duct region ID of 65531
+- A second duct region ID of 65532
```

Comment:
After discussing this with @eshemon and the rest of the ANL meshing team, the decision to hard code these values was made in order to provide visibility to the user to be able to easily map block and sideset names to their corresponding id's. We agree that this could result in unintentional collisions. Your proposal of a `get_unique_id` could certainly benefit these mesh generators in some way. However, using something like `get_unique_id`  could also make it more difficult for users to figure out any id remappings that occurred under the hood during the mesh generation stage. For most users of this functionality who would be just using this pin/assembly/core system to build a reactor there shouldn't be any major issue. There is a tradeoff on how to handle both block ID and boundary ids (automation vs. predictability for user) and to complicate things, we are discovering that some codes having particular requirements on numbering, etc. We are working on understanding all the requirements for boundary / block numbering from different stakeholders, so we propose to defer this to be addressed in a future PR. Let us know if this is agreeable.

Reply:
I really like the way the `Communicator::get_unique_tag()` method worked out, but I agree that it's not appropriate here - that was for internal ids, not user-visible ids.  If the user has a specific number in mind, then changing that number would be misleading and a potential source of error.

But on the other hand... isn't this truncation feature *also* kind of pulling out the rug from under them?  If I ask for 65535- but I get 65530- instead, I'm going to be in trouble whether the root problem was a conflicting generator or a 16-bit data type.  Simply throwing an error and exiting would be safer than trying to continue in either case.

I don't think it should block this PR, but do we have a separate issue open yet to discuss it?  If so would you tag me, and if not would you open one?  I'm kind of surprised that people want to worry about numeric ids at all - for both boundaries and subdomains we have mechanisms for translating between ids and strings, and the "strings" side of that isomorphism is both easier to work with and harder to cause collisions with.

Reply:
I tried to give ideas for limiting the chances of number collisions, but ultimately Roy's right. This is misleading and a potential source of error. Cant you just make the starting ID a parameter?

Reply:
well parameter wise, procerdual generation is decided at the reactor mesh param level, so it s not great to have that param either

Reply:
Currenrtly, wont we have a conflict between assembly of ID 2 and assmbly of ID 20 ?

Reply:
I think you're right about the possibility of a conflict here between assembly ID 2 and 20. Is the consensus to have the starting value as an input parameter and make it up to the user to make sure there are no conflicts there? Or even an offset parameter that separates the number of block id's between assembly id's that are successively numbered

Reply:
I would rather the first option
But if you want to design a bijective mapping between those IDs that would work for me too

Reply:
Resolving this. 
The solution chosen is simply to store the relevant extra integer IDs for assembly/pin/ring